pylint/config.py

```diff
@@ -594,11 +594,69 @@ def parse(self, to_parse, config):
                     type_ = definition.get('type')
                     validator = VALIDATORS.get(type_, lambda x: x)
                     value = validator(value)
-                config.set_option(option, value)
+                configuration.set_option(option, value)
+
+    def parse(self, to_parse, config):
+        self._parser.read(to_parse)
+        self.apply_to_configuration(config)
 
     def write(self, stream=sys.stdout):
-        # TODO: Check if option descriptions are written out
-        self._parser.write(stream)
+        """
+        Write out config to stream.
+
+        Includes option help and options with default values as
+        comments.
+        """
+        # We can't reuse self._parser because it doesn't have the help
+        # comments. allow_no_value lets us add comments as keys.
+        write_parser = configparser.ConfigParser(
+            default_section='MASTER',
+            allow_no_value=True,
+        )
+
+        # This makes keys case sensitive, needed for preserving comment
+        # formatting.
+        write_parser.optionxform = str
+
+        configuration = Configuration()
+        self.apply_to_configuration(configuration)
+        config_set_args = self.make_set_args(
+            self._option_definitions,
+            configuration
+        )
+        for args in config_set_args:
+            section = args[0]
+            if section != 'MASTER' and not write_parser.has_section(section):
+                write_parser.add_section(section)
+            write_parser.set(*args)
+        write_parser.write(stream)
+
+    @staticmethod
+    def make_set_args(option_definitions, configuration):
+        """Make args for configparser.ConfigParser.set."""
+        for option, definition in option_definitions.items():
+            section = definition.get('group', 'MASTER').upper()
+            help_lines = [
+                '# {}'.format(line)
+                for line in textwrap.wrap(definition.get('help', ''))
+            ]
+            for help_line in help_lines:
+                # Calling set with only two args does not leave an =
+                # symbol at the end of the line.
+                yield (section, help_line)
+            default = definition.get('default')
+            option_type = definition.get('type', 'string')
+            value = getattr(configuration, option)
+            if option_type == 'csv':
+                str_value = ',\n'.join(value)
+            else:
+                str_value = UNVALIDATORS[option_type](value)
```

Comment:
UNVALIDATORS looks to be general purpose and used for CLI stuff, so wanted to leave newlines out of them.